core/IssueQueue.cpp

```diff
@@ -0,0 +1,228 @@
+  auto pipe_vector = pipe_exe_pipe_mapping_.find(tgt_pipe);
+  pipe_vector->second.emplace_back(exe_pipe);
+  pipes_.push_back(exe_pipe);
+}
+
+std::unordered_map<std::string, olympia::ExecutePipe *>
+IssueQueue::getExePipes() {
+  return exe_pipes_;
+}
+
+void IssueQueue::getInstsFromDispatch_(const InstPtr &ex_inst) {
+  sparta_assert(ex_inst->getStatus() == Inst::Status::DISPATCHED,
+                "Bad instruction status: " << ex_inst);
+  appendIssueQueue_(ex_inst);
+  handleOperandIssueCheck_(ex_inst);
+}
+
+void IssueQueue::handleOperandIssueCheck_(const InstPtr &ex_inst) {
+  // FIXME: Now every source operand should be ready
+  const auto &src_bits = ex_inst->getSrcRegisterBitMask(reg_file_);
+  if (scoreboard_views_[reg_file_]->isSet(src_bits)) {
+    // Insert at the end if we are doing in order issue or if the scheduler is
+    // empty
+    ILOG("Sending to issue queue " << ex_inst);
+    if (in_order_issue_ == true || ready_queue_.size() == 0) {
+      ready_queue_.emplace_back(ex_inst);
+    } else {
+      // Stick the instructions in a random position in the ready queue
+      uint64_t issue_pos = uint64_t(std::rand()) % ready_queue_.size();
+      if (issue_pos == ready_queue_.size() - 1) {
+        ready_queue_.emplace_back(ex_inst);
+      } else {
+        uint64_t pos = 0;
+        auto iter = ready_queue_.begin();
+        while (iter != ready_queue_.end()) {
+          if (pos == issue_pos) {
+            ready_queue_.insert(iter, ex_inst);
+            break;
+          }
+          ++iter;
+          ++pos;
+        }
+      }
+    }
+    ev_issue_ready_inst_.schedule(sparta::Clock::Cycle(0));
+  } else {
+    scoreboard_views_[reg_file_]->registerReadyCallback(
+        src_bits, ex_inst->getUniqueID(),
+        [this, ex_inst](const sparta::Scoreboard::RegisterBitMask &) {
+          this->handleOperandIssueCheck_(ex_inst);
+        });
+    ILOG("Instruction NOT ready: " << ex_inst << " Bits needed:"
+                                   << sparta::printBitSet(src_bits));
+  }
+}
+
+void IssueQueue::readyExeUnit_(const uint32_t &readyExe) {
+  /*
+  TODO:
+  have a map/mask to see which execution units are ready
+  signal port
+  */
+  if (!ready_queue_.empty()) {
+    ev_issue_ready_inst_.schedule(sparta::Clock::Cycle(0));
+  }
+}
+
+void IssueQueue::sendReadyInsts_() {
+  sparta_assert(
+      ready_queue_.size() <= scheduler_size_,
+      "ready queue greater than issue queue size: " << scheduler_size_);
+  for (auto inst_itr = ready_queue_.begin(); inst_itr != ready_queue_.end();) {
+    auto inst = *inst_itr;
+    auto &valid_exe_pipe = pipe_exe_pipe_mapping_[inst->getPipe()];
+    bool sent = false;
+    for (auto &exe_pipe : valid_exe_pipe) {
+      if (exe_pipe->canAccept()) {
+        exe_pipe->issueInst(inst);
+        inst_itr = ready_queue_.erase(inst_itr);
+        popIssueQueue_(inst);
+        out_scheduler_credits_.send(
+            1, 0); // send credit back to dispatch, we now have more room in IQ
```

Comment:
Shouldn't this code be in `popIssueQueue`?

Reply:
Resolved in [6b0e363](https://github.com/riscv-software-src/riscv-perf-model/pull/140/commits/6b0e3633c1b7331170afda8920f596f9fd16b233)